src/stake_pool.rs

```diff
@@ -1,46 +1,1318 @@
+
+/// Don't bother adjusting stake if less than this amount of lamports will be affected
+/// (must be >= MIN_STAKE_ACCOUNT_BALANCE)
+const MIN_STAKE_CHANGE_AMOUNT: u64 = MIN_STAKE_ACCOUNT_BALANCE;
+
+fn get_minimum_stake_balance_for_rent_exemption(
+    rpc_client: &RpcClient,
+) -> Result<u64, Box<dyn error::Error>> {
+    rpc_client
+        .get_minimum_balance_for_rent_exemption(mem::size_of::<StakeState>())
+        .map_err(|err| format!("Error fetching rent exemption: {}", err).into())
 }
 
-pub struct SplStakePool {
+/// Seed for the transient stake account used by the staker
+fn staker_transient_stake_address_seed(vote_address: Pubkey) -> String {
+    format!("T{}", vote_address)[..32].to_string()
+}
+
+/// Staker's transient stake account
+///
+/// When adding a new validator to the pool, we must create a validator stake
+/// account with the rent-exempt amount + MIN_STAKE_ACCOUNT_BALANCE, and delegate
+/// the account to the appropriate vote address.  Once the stake is active, we
+/// can add it to the pool.
+///
+/// Separately, the stake pool enforces that a new validator stake account must
+/// have exactly the rent-exempt amount + MIN_STAKE_ACCOUNT_BALANCE as its balance.
+/// Since the validator stake account was staked for an epoch, it earned rewards during
+/// that epoch, bringing it over the enforced lamport amount. The extra amount
+/// is split into a transient stake account defined by this function.
+fn staker_transient_stake_address(authorized_staker: Pubkey, vote_address: Pubkey) -> Pubkey {
+    Pubkey::create_with_seed(
+        &authorized_staker,
+        &staker_transient_stake_address_seed(vote_address),
+        &solana_stake_program::id(),
+    )
+    .unwrap()
+}
+
+#[derive(Debug)]
+pub struct StakePoolOMatic {
     authorized_staker: Keypair,
     baseline_stake_amount: u64,
-    pool_address: Pubkey,
-    validator_info: HashMap<Pubkey, ValidatorInfo>,
+    stake_pool_address: Pubkey,
+    stake_pool: StakePool,
+    validator_list: ValidatorList,
 }
 
 pub fn new(
-    _rpc_client: &RpcClient,
+    rpc_client: &RpcClient,
     authorized_staker: Keypair,
-    pool_address: Pubkey,
+    stake_pool_address: Pubkey,
     baseline_stake_amount: u64,
-) -> Result<SplStakePool, Box<dyn error::Error>> {
-    Ok(SplStakePool {
+) -> Result<StakePoolOMatic, Box<dyn error::Error>> {
+    if baseline_stake_amount < MIN_STAKE_CHANGE_AMOUNT {
+        return Err(format!(
+            "baseline stake amount too small: {}",
+            Sol(baseline_stake_amount)
+        )
+        .into());
+    }
+
+    let account_data = rpc_client.get_account_data(&stake_pool_address)?;
+    let stake_pool = StakePool::try_from_slice(account_data.as_slice())
+        .map_err(|err| format!("Invalid stake pool {}: {}", stake_pool_address, err))?;
+    let account_data = rpc_client.get_account_data(&stake_pool.validator_list)?;
+    let validator_list = try_from_slice_unchecked::<ValidatorList>(&account_data.as_slice())
```

Comment:
I was originally imagining we'd do exactly this here (read the validator list off chain) but I'm now thinking the list should be passed in, like the other implementations.  The business logic in `src/main.rs` will soon be managing the testnet/mainnet validator identities more gracefully than a hard-coded list and will tie into the new foundation delegation website.  Fortunately it looks like the implementation already supports adding/removing validators from the stake pool ValidatorList

Reply:
That's right -- this validator list is specific to the stake pool, and can safely live independently of the `validator_list` defined in this repo.  Since they won't always be perfectly in sync, the implementation reads the `desired_validator_stake` to figure out which validators to add / remove from the stake pool, so in that sense the validator list is passed in during `apply`, just like stake_pool_v0

Reply:
Perfect!